src/pykeen/models/uncertainty.py

```diff
@@ -0,0 +1,334 @@
+# -*- coding: utf-8 -*-
+
+"""
+Analyze uncertainty.
+
+Currently, all implemented approaches are based on Monte-Carlo dropout [gal2016]_.
+Monte-Carlo dropout relies on the model having dropout layers. While dropout usually is
+turned off for inference / evaluation mode, MC dropout leaves dropout enabled. Thereby,
+if we run the same prediction method $k$ times, we get $k$ different predictions.
+The variance of these predictions can be used as an approximation of uncertainty, where
+larger variance indicates higher uncertainty.
```

Comment:
I know I'm being picky with the wording, but when you write uncertainty, what are you referring to being uncertain about? The correctness of the prediction?
What's the actionable part? if the variance is high, then trust the results less?

What's the threshold for high variance, even? Or should users look at the distribution of all variances?

Reply:
> I know I'm being picky with the wording, but when you write uncertainty, what are you referring to being uncertain about? The correctness of the prediction?

Technically, it is only for the scores, rather than the link prediction (i.e., e.g., rank). I changed the example to use BCE, i.e., a classification loss. Now we train the model more on the triple classification task ("does (h, r, t) hold?") rather than the link prediction task ("what are likely t for (h, r, ?)"), and thus the individual scores more of interest.

> What's the actionable part? if the variance is high, then trust the results less?
What's the threshold for high variance, even? Or should users look at the distribution of all variances?

Afaik, the absolute score is hard to interpret (it is the variance of predicted logits...). However, we can use it to compare between different predictions and order them. There are also evaluation methods for uncertainty quantification (e.g. from https://github.com/uncertainty-toolbox/uncertainty-toolbox), which often look whether the error rate correlates with the predicted uncertainty, i.e., whether most errors occur for more samples with higher predicted uncertainty.